recipes/google-cloud-cpp-core/meta.yaml

```diff
@@ -0,0 +1,364 @@
+    - libgrpc
+    - libprotobuf
+    - openssl
+    - nlohmann_json
+    - zlib
+
+outputs:
+  - name: libgoogle-cloud-common
+    script: build-libgoogle-cloud.sh   # [not win]
+    script: build-libgoogle-cloud.bat  # [win]
+    build:
+      run_exports:
+        # patch versions guaranteed to be ABI-compatible, see
+        # https://github.com/googleapis/google-cloud-cpp/blob/main/doc/adr/2023-05-03-patch-releases-are-ABI-compatible.md
+        - {{ pin_subpackage("libgoogle-cloud-common", max_pin="x.x") }}
+    requirements:
+      build:
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
+        - cmake
+        - ninja
+        - libgrpc
+        - libprotobuf
+      host:
+        - libabseil
+        - libcurl
+        - libgrpc
+        - libprotobuf
+        - openssl
+        - nlohmann_json
+        - libcxx  # [osx]
+      run:
+        - libcxx  # [osx]
+      run_constrained:
+        - libgoogle-cloud {{ version }} *_{{ PKG_BUILDNUM }}
+    test:
+      {{ test_subpackage("common")|indent(6) }}
+
+  - name: libgoogle-cloud-common-devel
+    script: build-libgoogle-cloud.sh   # [not win]
+    script: build-libgoogle-cloud.bat  # [win]
+    build:
+      run_exports:
+        - {{ pin_subpackage("libgoogle-cloud-common", max_pin="x.x") }}
+    requirements:
+      build:
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
+        - cmake
+        - ninja
+        - libcurl
+        - libgrpc
+        - libprotobuf
+        - openssl
+        - nlohmann_json
+        # cannot load package without zlib because it's specified
+        # as required; if possible, this should be removed upstream
+        - zlib
+      host:
+        - {{ pin_subpackage("libgoogle-cloud-common", exact=True) }}
+        - libabseil
+        - libcurl
+        - libgrpc
+        - libprotobuf
+        - openssl
+        - nlohmann_json
+      run:
+        - {{ pin_subpackage("libgoogle-cloud-common", exact=True) }}
+        - libabseil
+        - libcurl
+        - libgrpc
+        - libprotobuf
+        - openssl
+        - nlohmann_json
+        # cannot load package without zlib because it's specified
+        # as required; if possible, this should be removed upstream
+        - zlib        
+        - libcxx  # [osx]
+    test:
+      commands:
+        - ./run_test_common.sh   # [not win]
+        - ./run_test_common.bat  # [win]
+      requires:
+        - {{ compiler('cxx') }}
+        - cmake
+        - ninja
+      files:
+        - run_test_common.sh
+        - run_test_common.bat
+        - tests/CMakeLists.txt
+        - tests/quickstart.cc
+
+  - name: libgoogle-cloud-oauth2
+    {{ standard_subpackage("oauth2")|indent(4) }}
+  - name: libgoogle-cloud-oauth2-devel
+    {{ standard_devel_subpackage("oauth2")|indent(4) }}
+  - name: libgoogle-cloud-bigtable
+    {{ standard_subpackage("bigtable")|indent(4) }}
+  - name: libgoogle-cloud-bigtable-devel
+    {{ standard_devel_subpackage("bigtable")|indent(4) }}
```

Comment:
So these macros are pretty unusual (I've never seen them used somewhere else), but I guess in this case it's actually pretty justifiable.

The only thing I'd consider is bringing things closer together in terms of definition-site & usage-site, i.e.

```
# define standard_subpackage macro
# all uses of standard_subpackage macro

# define standard_devel_subpackage macro
# all uses of standard_devel_subpackage macro

# define child_subpackage macro
# all uses of child_subpackage macro
```
etc.

Reply:
I tried that and got some build errors, probably because I did something wrong.  Any suggestions on how to troubleshoot this?

```
Traceback (most recent call last):
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 228, in <module>
    build_all(os.path.join(root_dir, "recipes"), args.arch)
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 118, in build_all
    build_folders(recipes_dir, folders, arch, channel_urls)
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 155, in build_folders
    G = construct_graph(recipes_dir, worker=worker, run='build',
  File "/home/conda/staged-recipes-copy/.ci_support/compute_build_graph.py", line 400, in construct_graph
    add_recipe_to_graph(recipe_dir, graph, run, worker, conda_resolve,
  File "/home/conda/staged-recipes-copy/.ci_support/compute_build_graph.py", line 219, in add_recipe_to_graph
    rendered = _get_or_render_metadata(recipe_dir, worker, config=config, finalize=finalize)
  File "/opt/conda/lib/python3.10/site-packages/conda/exports.py", line 168, in __call__
    return self.func(*args, **kw)
  File "/home/conda/staged-recipes-copy/.ci_support/compute_build_graph.py", line 210, in _get_or_render_metadata
    api.render(meta_file_or_recipe_dir, platform=platform, arch=arch,
  File "/opt/conda/lib/python3.10/site-packages/conda_build/api.py", line 50, in render
    metadata_tuples = render_recipe(
  File "/opt/conda/lib/python3.10/site-packages/conda_build/render.py", line 969, in render_recipe
    m = MetaData(recipe_dir, config=config)
  File "/opt/conda/lib/python3.10/site-packages/conda_build/metadata.py", line 1111, in __init__
    self.parse_again(permit_undefined_jinja=True, allow_no_other_outputs=True)
  File "/opt/conda/lib/python3.10/site-packages/conda_build/metadata.py", line 1203, in parse_again
    self._get_contents(
  File "/opt/conda/lib/python3.10/site-packages/conda_build/metadata.py", line 1916, in _get_contents
    rendered = template.render(environment=env)
  File "/opt/conda/lib/python3.10/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/opt/conda/lib/python3.10/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/home/conda/staged-recipes-copy/recipes/google-cloud-cpp-core/meta.yaml", line 60, in top-level template code
    - libprotobuf
  File "/opt/conda/lib/python3.10/site-packages/jinja2/filters.py", line 818, in do_indent
    lines = s.splitlines()
TypeError: 'NoneType' object is not callable
valid configs are {'linux64_cuda112', 'linux64', 'osx64', 'win64'}
Using linux64 configuration
Traceback (most recent call last):
  File "/usr/local/google/home/coryan/staged-recipes/build-locally.py", line 101, in <module>
    main()
  File "/usr/local/google/home/coryan/staged-recipes/build-locally.py", line 95, in main
    run_docker_build(ns)
  File "/usr/local/google/home/coryan/staged-recipes/build-locally.py", line 33, in run_docker_build
    subprocess.check_call([script])
  File "/usr/lib/python3.11/subprocess.py", line 413, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['.scripts/run_docker_build.sh']' returned non-zero exit status 1.
```